src/absil/illib.fs

```diff
@@ -432,6 +436,48 @@ module List =
     let existsSquared f xss = xss |> List.exists (fun xs -> xs |> List.exists (fun x -> f x))
     let mapiFoldSquared f z xss =  mapFoldSquared f z (xss |> mapiSquared (fun i j x -> (i,j,x)))
 
+module ResizeArray =
+
+    /// Split a ResizeArray into an array of smaller chunks.
+    /// This requires `items/chunkSize` Array copies of length `chunkSize` if `items/chunkSize % 0 = 0`,
+    /// otherwise `items/chunkSize + 1` Array copies.
+    let chunkBySize chunkSize f (items: ResizeArray<'t>) =
+        // we could use Seq.chunkBySize here, but that would involve many enumerator.MoveNext() calls that we can sidestep with a bit of math
+        let itemCount = items.Count
+        if itemCount = 0
+        then [||]
+        else
+            let chunksCount =
+                match itemCount / chunkSize with
+                | n when itemCount % chunkSize = 0 -> n
+                | n -> n + 1 // any remainder means we need an additional chunk to store it
+
+            [| for index in 0..chunksCount-1 do
+                let startIndex = index * chunkSize
+                let takeCount = min (itemCount - startIndex) chunkSize
+
+                let holder = Array.zeroCreate takeCount
+                // we take a bounds-check hit here on each access.
+                // other alternatives here include
+                // * iterating across an IEnumerator (incurs MoveNext penalty)
+                // * doing a block copy using `List.CopyTo(index, array, index, count)` (requires more copies to do the mapping)
+                // none are significantly better.
+                for i in 0 .. takeCount - 1 do
+                    holder.[i] <- f items.[i]
+                yield holder |]
+
+    /// Split a large ResizeArray into a series of array chunks that are each under the
+    /// Large Object Heap limit, in order to prevent the entire array from not being garbage-collected.
```

Comment:
> in order to prevent the entire array from not being garbage-collected.

The whole thing can still be garbage-collected, but the distinction is that it can be collected in chunks and in the background rather than having the GC pause all threads. Even then, it's possible that a stop-the-world collection could happen, but ensuring that things don't end up on the LOH makes that probability lower.

Reply:
That's a good and subtle point. I updated the comment to help head off confusion in that regard.